miri-script/src/commands.rs

```diff
@@ -0,0 +1,629 @@
+        let cargo_miri_dir = path!(self.miri_dir / "cargo-miri");
+        self.install_to_dir(&sh, cargo_miri_dir, flags)?;
+        Ok(())
+    }
+
+    fn build(&self, flags: &[OsString]) -> Result<()> {
+        // Build, and let caller control flags.
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        Self::build_package(
+            miri_manifest.as_ref(),
+            &self.active_toolchain,
+            &self.cargo_extra_flags,
+            flags,
+        )?;
+        Self::build_package(
+            cargo_miri_manifest.as_ref(),
+            &self.active_toolchain,
+            &self.cargo_extra_flags,
+            flags,
+        )?;
+        Ok(())
+    }
+
+    fn check(&self, flags: &[OsString]) -> Result<()> {
+        fn check_package(
+            // Path to Cargo.toml file of a package to check.
+            path: &OsStr,
+            toolchain: impl AsRef<OsStr>,
+            extra_flags: &[String],
+            all_targets: bool,
+            args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+        ) -> Result<()> {
+            let all_targets: Option<&OsStr> = all_targets.then_some("--all-targets".as_ref());
+            let sh = Shell::new()?;
+            cmd!(sh, "cargo +{toolchain} check {extra_flags...} --manifest-path {path} {all_targets...} {args...}").run()?;
+            Ok(())
+        }
+        // Check, and let caller control flags.
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        check_package(
+            miri_manifest.as_ref(),
+            &self.active_toolchain,
+            &self.cargo_extra_flags,
+            true,
+            flags,
+        )?;
+        check_package(
+            cargo_miri_manifest.as_ref(),
+            &self.active_toolchain,
+            &self.cargo_extra_flags,
+            false,
+            flags,
+        )?;
+        Ok(())
+    }
+
+    fn test(&mut self, bless: bool, flags: &[OsString]) -> Result<()> {
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        // First build and get a sysroot.
+        Self::build_package(
+            miri_manifest.as_ref(),
+            &self.active_toolchain,
+            &self.cargo_extra_flags,
+            std::iter::empty::<OsString>(),
+        )?;
+        self.find_miri_sysroot()?;
+        let extra_flags = &self.cargo_extra_flags;
+        // Then test, and let caller control flags.
+        // Only in root project as `cargo-miri` has no tests.
+        let sh = self.shell()?;
+        if bless {
+            sh.set_var("MIRI_BLESS", "Gesundheit");
+        }
+        let toolchain: &OsStr = self.active_toolchain.as_ref();
+        cmd!(
+            sh,
+            "cargo +{toolchain} test {extra_flags...} --manifest-path {miri_manifest} -- {flags...}"
```

Comment:
The `-- {flags..}` here is what got CI to work despite the parser issue, but it breaks other things -- for instance, `./miri test --all-features` should pass that flag to cargo (before the `--`), not to the test runner. (Same for clippy.)